### PR TITLE
Fix categorical_crossentropy label smoothing for custom axis

### LINKED ISSUE #23451: Categorical_crossentropy and categorical_focal_crossentropy ignore axis for smoothing and warnings
When using categorical_crossentropy or categorical_focal_crossentropy with a custom axis (e.g., in 3D tensors with (batch, classes, spatial) format), the label_smoothing logic and the shape-related SyntaxWarning both ignore the user-provided axis and default to the last dimension (-1).

### Steps to Reproduce
```
import numpy as np
from keras.src.losses import losses

# 3D tensor: (batch, classes, spatial) -> axis=1 has 3 classes, axis=2 has 2 spatial dims
y_true = np.array([[[1, 0], [0, 1], [0, 0]]], dtype="float32") # Shape (1, 3, 2)
y_pred = np.array([[[0.8, 0.1], [0.1, 0.8], [0.1, 0.1]]], dtype="float32")

# We specify axis=1, but smoothing will incorrectly use axis=-1 (length 2)
loss = losses.categorical_crossentropy(y_true, y_pred, label_smoothing=0.1, axis=1)
```

Colab notebook with bug reproduction: https://colab.research.google.com/drive/1LyDEp9HJI-MoCOAwTfmqbEBV8ec-O9cj?resourcekey=0-IuR4p5aMkSy2FYyeEgYC_g&usp=sharing

  ### Expected Behavior
   1. Label Smoothing: Should calculate num_classes based on the specified axis.
   2. Warnings: The SyntaxWarning (for single-class inputs) should check the class counts on the specified axis, not just the last dimension.
   3. Robustness: The code should canonicalize the axis to handle negative indices and verify static shapes during tracing.

  ### Actual Behavior
  The implementation hardcodes shape[-1] for both smoothing math and warnings, leading to incorrect loss values when axis is not -1.

### PR DESCRIPTION
**Pull request by buildwithsuhana**
Fixes https://github.com/keras-team/keras/issues/23451

## Description
This PR fixes a bug in categorical_crossentropy and categorical_focal_crossentropy where the label_smoothing logic incorrectly assumed the class dimension was always the last axis.

Root Cause
When label_smoothing is enabled, the loss function calculates the number of classes (num_classes) to redistribute probability mass from the ground truth. Previously, this was hardcoded to num_classes = ops.shape(y_true)[-1].

If a user specified a custom axis (e.g., axis=1 for (batch, classes, spatial) data), the implementation would use the size of the last dimension (often a spatial or temporal dimension) instead of the actual class dimension. This resulted in:
   1. Invalid probability distributions: The smoothed labels would not sum to 1.0.
   2. Incorrect loss values: Model training would be driven by silently corrupted gradients.

Additionally, the implementation contained a warning for singleton class dimensions that was also hardcoded to check only the last axis.

Reproduction script: https://colab.research.google.com/drive/1LyDEp9HJI-MoCOAwTfmqbEBV8ec-O9cj?resourcekey=0-IuR4p5aMkSy2FYyeEgYC_g&usp=sharing

## Contributor Agreement

**Please review our [AI-Assisted Contribution Policy](../CONTRIBUTING.md#ai-assisted-contribution-policy) and check all boxes below before submitting your PR for review:**

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

**Note:** Failing to adhere to this agreement may result in your future PRs no longer being reviewed.

### CODE REVIEW

**Comment by jeffcarp** [comment]
Is there a bug associated with this?

File: keras/src/losses/losses.py [hunk 1]
```diff
@@ -2184,7 +2184,10 @@ def categorical_crossentropy(
     y_pred = ops.convert_to_tensor(y_pred)
     y_true = ops.cast(y_true, y_pred.dtype)
 
-    if y_pred.shape[-1] == 1:
+    if y_pred.shape is not None:
+        axis = canonicalize_axis(axis, len(y_pred.shape))
+
+    if y_pred.shape is not None and y_pred.shape[axis] == 1:
```
Comment (jeffcarp) [inline keras/src/losses/losses.py]: This can just be a sub-condition of the above if statement

File: keras/src/losses/losses_test.py [hunk 2]
```diff
@@ -1165,6 +1165,15 @@ def test_scalar_weighted(self):
...
         self.assertAlmostEqual(loss, 0.1317)
 
+        # Test warning for invalid shape.
+        y_true = np.array([[1], [1]], dtype="float32")
+        y_pred = np.array([[0.5], [0.5]], dtype="float32")
+        with pytest.warns(
+            SyntaxWarning,
+            match=r"expected y_pred.shape to be \(batch_size, num_classes\)",
+        ):
+            losses.categorical_crossentropy(y_true, y_pred)
```
Comment (jeffcarp) [inline keras/src/losses/losses_test.py:1177]: Can you make this its own test method? (here and below)

Comment (jeffcarp) [reply keras/src/losses/losses_test.py:1177]: I think this test would have already passed before this PR, maybe test with 3D shapes?

- Shape (2, 1, 4) with axis=1 → should warn (1 class)
- Shape (2, 4, 1) with axis=1 → should NOT warn (4 classes, even though trailing dim is 1)

(here and below)

**Review by jeffcarp — CHANGES_REQUESTED** [review]

File: keras/src/losses/losses.py (hunk 1 above)
Comment (buildwithsuhana) [reply keras/src/losses/losses.py]: Done

File: keras/src/losses/losses_test.py (hunk 2 above)
Comment (buildwithsuhana) [reply keras/src/losses/losses_test.py:1177]: Done

**Comment by buildwithsuhana** [comment]
> Is there a bug associated with this?

Raised this issue for the bug: https://github.com/keras-team/keras/issues/23451

**Review by jeffcarp — APPROVED** [review]
Thanks!
